### PR TITLE
Enhancements to table batch operations

### LINKED ISSUE #14583: Provide better response parsing for batch operations
Currently batch operation responses return an arrray of `Response` objects without any deserialized value or explicit parsing of known headers such as `Location`, which identifies an inserted or updated entity.

Consider providing better response parsing and correlation between batch sub-request and batch sub-response.

### LINKED ISSUE #15307: Azure.Data.Tables Readme issue
1.
Section [link](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/tables/Azure.Data.Tables#creating-a-cosmos-db)

image

Reason: 
Azure Cosmos DB database with Azure Table API specified.

Suggestion: 
 Update  to 
 ```
az cosmosdb create --name MyCosmosDBDatabaseAccount --capabilities EnableTable --resource-group MyResourceGroup --subscription MySubscription 
```

@jongio for notification.

### PR DESCRIPTION
**Pull request by christothes**
The focus of these changes is to make some enhancements to table batch operations and fix a minor TableEntity property issue.

- `TablesBatch` renamed to `TablesTransactionalBatch`
  - This is to distinguish it from the broader "streaming batch" scenario
- Multi-part responses are now configurable with regard to expecting CRLF to separate boundaries.
  - This is to accommodate Cosmos response bodies whic are inconsistent with CRLF and LF usage
- Fixed an issue where `TableEntity.TimeStamp` was always null due to an invalid cast
  - We were also not casting the service `TimeStamp` property correctly (treating it as a DateTime rather than string)
- Re-enabled Cosmos batch operation tests as they now are able to pass

closes #14583 
fixes #15307

### CODE REVIEW

File: sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs [hunk 1]
```diff
@@ -144,13 +172,23 @@ public virtual async Task<Response<List<Response>>> SubmitBatchAsync(Cancellatio
         /// </summary>
         /// <typeparam name="T"></typeparam>
         /// <returns></returns>
-        public virtual Response<List<Response>> SubmitBatch<T>(CancellationToken cancellationToken = default)
+        public virtual Response<TablesBatchResponse> SubmitBatch<T>(CancellationToken cancellationToken = default)
```
Comment (tg-msft) [inline sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs:175]: Nit - consider using the shared implementation pattern where you pass `bool async` to determine if you're executing synchronously or not.  This code may grow more complex over time.

File: sdk/tables/Azure.Data.Tables/src/Extensions/DictionaryTableExtensions.cs [hunk 2]
```diff
@@ -104,7 +104,7 @@ internal static void CastAndRemoveAnnotations(this IDictionary<string, object> e
                 entity[annotation] = typeAnnotationsWithKeys[annotation].typeAnnotation switch
                 {
                     TableConstants.Odata.EdmBinary => Convert.FromBase64String(entity[annotation] as string),
-                    TableConstants.Odata.EdmDateTime => DateTime.Parse(entity[annotation] as string, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind),
+                    TableConstants.Odata.EdmDateTime => DateTimeOffset.Parse(entity[annotation] as string, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind),
```
Comment (tg-msft) [inline sdk/tables/Azure.Data.Tables/src/Extensions/DictionaryTableExtensions.cs:107]: Nit - is this visible to users?  Would it need a Changelog entry?

**Review by tg-msft — APPROVED** [review]

File: sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs [hunk 3]
```diff
@@ -122,6 +123,16 @@ public static partial class TablesModelFactory
     {
         public static Azure.Data.Tables.Models.TableItem TableItem(string tableName, string odataType, string odataId, string odataEditLink) { throw null; }
     }
+    public partial class TablesTransactionalBatch
```
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs:126]: I agree, we should probably use the singular consistently

File: sdk/tables/Azure.Data.Tables/src/Extensions/DictionaryTableExtensions.cs (hunk 2 above)
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/Extensions/DictionaryTableExtensions.cs:107]: We were ultimately casting to `DateTimeOffset` in `TableEntity` so this is just an internal implementation detail and a bug fix. Previously, the Timestamp property always returned null.

File: sdk/tables/Azure.Data.Tables/src/TableRestClient.cs [hunk 4]
```diff
@@ -91,9 +70,15 @@ public async Task<Response<List<Response>>> SendBatchRequestAsync(HttpMessage me
                         var responses = await Multipart.ParseAsync(
                             message.Response.ContentStream,
                             message.Response.Headers.ContentType,
+                            false,
                             true,
                             cancellationToken).ConfigureAwait(false);
 
+                        if (responses.Length == 1 && responses.Any(r => r.Status >= 400))
```
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableRestClient.cs:77]: Good question. I don’t think so, but I can’t say for sure. Will verify.

File: sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs [hunk 5]
```diff
@@ -104,33 +111,54 @@ public virtual void UpdateEntity<T>(T entity, ETag ifMatch, TableUpdateMode mode
                 ifMatch.ToString(),
                 tableEntityProperties: entity.ToOdataAnnotatedDictionary(),
                 queryOptions: new QueryOptions() { Format = _format });
+
+            _requestMessages.Enqueue((entity.RowKey, message));
+            _requestLookup[entity.RowKey] = (message, RequestType.Update);
```
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs:116]: It’s not allowed by the service, but good catch that we wouldn’t handle that properly here.

File: sdk/tables/Azure.Data.Tables/src/TableRestClient.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TableRestClient.cs:77]: 304 would be the only reasonable 3xx response, but none of the batch operations should produce that. Even if it did, I think we'd handle it correctly here as none of the other response types produce content either, as currently configured.

File: sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs (hunk 1 above)
Comment (christothes) [reply sdk/tables/Azure.Data.Tables/src/TablesTransactionalBatch.cs:175]: I'll defer until it does get more complex.

**Pull request merged**